### ISSUE TITLE
Weekly event returned on wrong dates

### ISSUE THREAD

**Issue by jgrygierek** [u01jmg3/ics-parser#105]
Hi, I have the same problem as described here:

RE: #100

```
BEGIN:VCALENDAR
PRODID:-//Google Inc//Google Calendar 70.9054//EN
VERSION:2.0
CALSCALE:GREGORIAN
METHOD:PUBLISH
X-WR-CALNAME:Praca
X-WR-TIMEZONE:Europe/Warsaw
X-WR-CALDESC:test@test.pl
BEGIN:VTIMEZONE
TZID:Europe/Warsaw
X-LIC-LOCATION:Europe/Warsaw
BEGIN:DAYLIGHT
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
TZNAME:CEST
DTSTART:19700329T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=-1SU
END:DAYLIGHT
BEGIN:STANDARD
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
TZNAME:CET
DTSTART:19701025T030000
RRULE:FREQ=YEARLY;BYMONTH=10;BYDAY=-1SU
END:STANDARD
END:VTIMEZONE
BEGIN:VEVENT
DTSTART;TZID=Europe/Warsaw:20170309T000000
DTEND;TZID=Europe/Warsaw:20170310T000000
RRULE:FREQ=WEEKLY;BYDAY=TH
DTSTAMP:20170309T095228Z
UID:7mb54fkuh2v93tc8lpkp3m7puk@google.com
CREATED:20170309T093023Z
DESCRIPTION:
LAST-MODIFIED:20170309T094007Z
LOCATION:
SEQUENCE:1
STATUS:CONFIRMED
SUMMARY:test2
TRANSP:OPAQUE
END:VEVENT
BEGIN:VEVENT
DTSTART;VALUE=DATE:20170307
DTEND;VALUE=DATE:20170308
RRULE:FREQ=WEEKLY;BYDAY=TU
DTSTAMP:20170309T095228Z
UID:iqhokmdsti6227spvkb4edaqao@google.com
CREATED:20170306T095043Z
DESCRIPTION:
LAST-MODIFIED:20170309T090512Z
LOCATION:
SEQUENCE:1
STATUS:CONFIRMED
SUMMARY:test
TRANSP:TRANSPARENT
END:VEVENT
END:VCALENDAR
```

**Comment by u01jmg3** [comment]
Can you be more specific about what your issue is so I don't have to try and guess?

Your iCal snippet consists of 2 events:
- The `test` event has a rule stating that it repeats **every Tuesday** and the parse correctly outputs this
  1. `test (07-03-2017 00:00)`
  1. `test (14-03-2017 00:00)`
  1. `test (21-03-2017 00:00)`
  1. `test (28-03-2017 00:00)`
  1. `test (04-04-2017 00:00)`
  1. `...`
- The `test2` event has a rule stating that it repeats **every Thursday** and the parse correctly outputs this
  1. `test2 (09-03-2017 00:00)`
  1. `test2 (16-03-2017 00:00)`
  1. `test2 (23-03-2017 00:00)`
  1. `test2 (30-03-2017 00:00)`
  1. `test2 (06-04-2017 00:00)`
  1. `...`

I therefore cannot see what is wrong without more explanation.

**Comment by jgrygierek** [comment]
When I run this parser, each recursive event is shifted by one day. Original events are ok: 07.03 and 09.03. But next events have dates: 08.03, 15.03, 16.03, ... and 10.03, 17.03, ...

**Comment by u01jmg3** [comment]
You are still going to have to be more specific. The examples I list include the recursive event fired by the rule so I'm struggling to know where your issue lies.

If you can state exactly what code you are running against what specific iCal event and let me know what outcome you are seeing versus what you expect to see. This way I will be able to help you.

**Comment by jgrygierek** [comment]
//latest version of ics parser, without changing anything by me (original settings)
$ical = new \ICal\ICal(Yii::getAlias('@web').'ics/basic.ics');
var_dump($ical->events());

Part of results:

```
object(ICal\EventObject)#56 (30) {
  ["summary"]=>
  string(5) "test2"
  ["dtstart"]=>
  string(15) "20170309T000000"
  ["dtend"]=>
  string(15) "20170310T000000"
 [..]
  ["dtstart_array"]=>
  array(3) {
    [0]=>
    array(1) {
      ["TZID"]=>
      string(13) "Europe/Warsaw"
    }
    [1]=>
    string(15) "20170309T000000"
    [2]=>
    int(1489014000)
  }
  ["dtend_array"]=>
  array(3) {
    [0]=>
    array(1) {
      ["TZID"]=>
      string(13) "Europe/Warsaw"
    }
    [1]=>
    string(15) "20170310T000000"
    [2]=>
    int(1489100400)
  }
  ["rrule_array"]=>
  array(2) {
    [0]=>
    array(0) {
    }
    [1]=>
    string(20) "FREQ=WEEKLY;BYDAY=TH"
  }
  ["rrule"]=>
  string(20) "FREQ=WEEKLY;BYDAY=TH"
 [..]
  ["dtstart_tz"]=>
  string(15) "20170309T010000"
  ["dtend_tz"]=>
  string(15) "20170310T010000"
}

object(ICal\EventObject)#112 (31) {
  ["summary"]=>
  string(5) "test2"
  ["dtstart"]=>
  string(15) "20170310T010000"
  ["dtend"]=>
  string(15) "20170311T010000"
  [..]
  ["dtstart_array"]=>
  array(3) {
    [0]=>
    array(1) {
      ["TZID"]=>
      string(13) "Europe/Warsaw"
    }
    [1]=>
    string(15) "20170310T010000"
    [2]=>
    int(1489104000)
  }
  ["dtend_array"]=>
  array(3) {
    [0]=>
    array(1) {
      ["TZID"]=>
      string(13) "Europe/Warsaw"
    }
    [1]=>
    string(15) "20170311T010000"
    [2]=>
    int(1489190400)
  }
  ["rrule_array"]=>
  array(2) {
    [0]=>
    array(0) {
    }
    [1]=>
    string(20) "FREQ=WEEKLY;BYDAY=TH"
  }
  ["rrule"]=>
  string(20) "FREQ=WEEKLY;BYDAY=TH"
  [...]
  ["dtstart_tz"]=>
  string(15) "20170310T020000"
  ["dtend_tz"]=>
  string(15) "20170311T020000"
}
```

Recursive event is moved by one day.

After adding events to the full calendar (dtstart and dtend), the result is:
![screen](https://cloud.githubusercontent.com/assets/3065409/23786877/a0c4510a-056f-11e7-9e31-a249ba393ed4.PNG)


If I change $useTimeZoneWithRRules to true, time of recursive events is changed to 00:00 - 00:00, but events still are moved by 1 day.

**Comment by acmurmeltier69** [comment]
Hi
Same problem here
Calendar event is;

> BEGIN:VCALENDAR
PRODID:-//sample MIMEDIR//EN
VERSION:2.0
METHOD:PUBLISH
BEGIN:VEVENT
DTSTART;VALUE=DATE:20170317
DTEND;VALUE=DATE:20170318
DESCRIPTION:Sample Event
LOCATION:Germany
RRULE:FREQ=WEEKLY;UNTIL=20171224T000000Z;INTERVAL=2;BYDAY=FR;WKST=MO
TRANSP:TRANSPARENT
SEQUENCE:0
UID:14885238391@abc.de
DTSTAMP:20170303T065039Z
SUMMARY:Sample Event
CLASS:PUBLIC
CREATED:20170222T151357Z
LAST-MODIFIED:20170222T151357Z
END:VEVENT
END:VCALENDAR

Expected dates: (raw format)
* 20170317 
* 20170331
* 20170417
...
Got:
* 2017-03-17
* 20170318T010000Z
* 20170401T010000Z
* 20170415T010000Z
...

So the first recurrence after DTSTART is one day after DTSTART instead of the specified intervall (in my example 2 weeks later)
This seems to be the same problem as described above.
It would be great if someone could fix that

**Comment by u01jmg3** [comment]
The rule (`FREQ=WEEKLY;UNTIL=20171224T000000Z;INTERVAL=2;BYDAY=FR;WKST=MO`) from from the event you have provided can be summarised as `every 2 weeks on Friday until December 24, 2017`.

The parser correctly outputs:

- Sample Event (17-03-2017 00:00)
- Sample Event (31-03-2017 00:00)
- Sample Event (14-04-2017 00:00)
- Sample Event (28-04-2017 00:00)
- ...

The expected 3rd date you quote, `20170417`, is not correct and instead the parser correctly produces `20170414` as the rule states the event falls on a Friday. `20170417` falls on a Monday which is not what your rule is stating.

Also the dates you quote as getting do not match with what the latest version of the parser is producing. Here is output running the latest version of the parser against your single event.

<img width="795" alt="demo" src="https://cloud.githubusercontent.com/assets/1266205/24071979/4682105c-0bd6-11e7-8211-c83a37011140.png">

**Comment by u01jmg3** [comment]
@jgrygierek

With `$useTimeZoneWithRRules` set to `false`, this is the output I get for your two events `test` and `test2` which is correct.

- `test`:
  - (No mention of `08-03-2017` like your screenshot)

<img width="795" alt="test" src="https://cloud.githubusercontent.com/assets/1266205/24072082/96741478-0bd8-11e7-857b-c5ef6039a12e.png">

- `test2`:
  - (No mention of `10-03-2017` like your screenshot)

<img width="795" alt="test2" src="https://cloud.githubusercontent.com/assets/1266205/24072083/987a51d8-0bd8-11e7-96ca-38d4c57f3393.png">

**Comment by u01jmg3** [comment]
@acmurmeltier69

The rrule (`FREQ=WEEKLY;UNTIL=20171224T000000Z;INTERVAL=2;BYDAY=FR;WKST=MO`) from from the event you have provided can be summarised as `every 2 weeks on Friday until December 24, 2017`.

The parser correctly outputs:

- Sample Event (17-03-2017 00:00)
- Sample Event (31-03-2017 00:00)
- Sample Event (14-04-2017 00:00)
- Sample Event (28-04-2017 00:00)
- ...

The expected 3rd date you quote, `20170417`, is not correct and instead the parser correctly produces `20170414` as the rule states the event falls on a Friday. `20170417` falls on a Monday which is not what your rule is stating.

Also the dates you quote as getting do not match with what the latest version of the parser is producing. Here is output running the latest version of the parser against your single event.

<img width="795" alt="demo" src="https://cloud.githubusercontent.com/assets/1266205/24071979/4682105c-0bd6-11e7-8211-c83a37011140.png">

**Comment by acmurmeltier69** [comment]
@u01jmg3: Thanks for the quick reply. 
You are right that the third date in my example should be 20170414. That was a typo on my side. Sorry!

I just downloaded the lazest release (v2.02) again to be sure and still get:
![image](https://cloud.githubusercontent.com/assets/25886295/24072368/be2c5f58-0be6-11e7-873e-0f260dfdf10a.png)

Any idea?
P.S: I am running the script on a STRATO Server where I cannot use composer and autoload.
Therefore I had to install the source manually and add 
'include "EventObject.php";' in line 14 of ICal.php and 
include "ICal/ICal.php"; into the example file.
I don't assume that is what causes the issue

**Comment by acmurmeltier69** [comment]
strange indeed.
I am using php 5.7 but also tried 7.0 (these are the only options I get).
I have put the demo online at  http://www.mnbvcx.eu/codetest/icaltest/index.php
The complete set of files is attached.
Can you send me a zip with your test package and I can try to run that on my server?
[icaltest.zip](https://github.com/u01jmg3/ics-parser/files/852532/icaltest.zip)

Sorry for the effort. It's probably a stupid PHP error on my side as I am still a newbie

**Comment by u01jmg3** [comment]
- I have just run your attached code and it works as expected on my setup; I've tested it against PHP `5.3.8`, `5.6.19` and `7.0.4` and all okay.
- The only thing I can think of that could be different is your timezone settings (`date_default_timezone_get()`).
- Without being able to debug in an environment that causes the error I can't really do anything on my end.
  - Is access available to your server so I can debug?
  - Unless you can do some debugging?

**Comment by acmurmeltier69** [comment]
Hi
it may indeed be a timezone thing.
I just installed a local apache with PHP5.6.30 and also get the wrong result.

Would it help if I set up a folder where you can upload files via ftp?
send me an IM to acmurmeltier69 AT mnbvcx DOT EU

**Comment by u01jmg3** [comment]
Maintenance release `v2.0.3` will address the timezone issues seen in this issue

**Comment by u01jmg3** [comment]
Fixed by the release of [`v2.0.3`](https://github.com/u01jmg3/ics-parser/releases/tag/v2.0.3)

**Comment by acmurmeltier69** [comment]
Hi Jonny
I found another (or two?) issues in the ICS parser. https://github.com/u01jmg3/ics-parser/issues/113 
Not sure if these are time--zone related but if you want to test in German timezone, the ftp account is still active.
I started a bt of debugging.The first issue (incorrect recurrance dates if WKST=MO) may have to do with the way the weeks are iterated
On the seconf part (exdate), I think the problem is in the multi-line concatenation around line 385.At least in my debug utputs, the continuation lines seem to be added to the line before the line that has the EXDATE statement.
Andi

 

    On Monday, March 20, 2017 12:48 PM, Jonathan Goode <notifications@github.com> wrote:
 

 Closed #105.—

**Comment by u01jmg3** [comment]
Shall take a look tomorrow - thanks
